### PR TITLE
added ANTLR visitor for parsing Logstash configuration

### LINKED ISSUE #465: Implement Logstash Configuration Parsing
**Is your feature request related to a problem? Please describe.**

This is part of the Logstash Configuration support. We need to parse Logstash Configuration files.

**Describe the solution you'd like**

Parse Logstash Configuration files using ANTLR.

### PR DESCRIPTION
**Pull request by asifsmohammed**
### Description
- Added Logstash visitor which populates model POJO's
- Added Exception classes for converter
- Updated checkstyle file to suppress checks for ANTLR generated files

 
### Issues Resolved
Resolve #465 
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [ ] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/exception/LogstashConfigurationException.java [hunk 1]
```diff
@@ -0,0 +1,14 @@
+package org.opensearch.dataprepper.logstash.exception;
+
+/**
+ * Exception for Logtsash converter
```
Comment (sshivanii) [inline data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/exception/LogstashConfigurationException.java:4]: typo in `Logstash`

File: data-prepper-logstash-configuration/src/main/antlr/Logstash.g4 [hunk 2]
```diff
@@ -0,0 +1,138 @@
+grammar Logstash;
```
Comment (sshivanii) [inline data-prepper-logstash-configuration/src/main/antlr/Logstash.g4:1]: Please include Java docs for this grammar file and add `@since 1.2`

File: data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/exception/LogstashParsingException.java [hunk 3]
```diff
@@ -0,0 +1,14 @@
+package org.opensearch.dataprepper.logstash.exception;
+
+/**
+ * Exception for visitor when it's unable to convert into Logstash models
```
Comment (sshivanii) [inline data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/exception/LogstashParsingException.java:4]: I'd like the javadoc to be more descriptive about what is Visitor? You can update this to `LogstashVisitor` and in that mention what the visitor class does.

File: data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/parser/LogstashVisitor.java [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+
+@SuppressWarnings("rawtypes")
+public class LogstashVisitor extends LogstashBaseVisitor {
+
+    private List<LogstashPlugin> logstashPluginList;
+    private final Map<LogstashPluginType, List<LogstashPlugin>> pluginSections = new LinkedHashMap<>();
+    private final Map<String, Object> hashEntries = new LinkedHashMap<>();
+
+    @Override
+    public Object visitConfig(LogstashParser.ConfigContext ctx) {
```
Comment (sshivanii) [inline data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/parser/LogstashVisitor.java:33]: can we rename this to `LogstashParser.ConfigContext configContext`?

File: data-prepper-logstash-configuration/src/test/java/org/opensearch/dataprepper/logstash/parser/TestDataProvider.java [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+
+
+public class TestDataProvider {
+    static final String RANDOM_STRING_1 = UUID.randomUUID().toString();
+    static final String RANDOM_STRING_2 = UUID.randomUUID().toString();
+    static final String RANDOM_VALUE = String.valueOf(new Random().nextInt(1000));
+
+    // data
```
Comment (sshivanii) [inline data-prepper-logstash-configuration/src/test/java/org/opensearch/dataprepper/logstash/parser/TestDataProvider.java:26]: comment not needed here

File: data-prepper-logstash-configuration/src/main/antlr/Logstash.g4 (hunk 2 above)
Comment (asifsmohammed) [reply data-prepper-logstash-configuration/src/main/antlr/Logstash.g4:1]: I think adding a java doc for files other than Java classes is not useful.

File: data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/parser/LogstashVisitor.java (hunk 4 above)
Comment (asifsmohammed) [reply data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/parser/LogstashVisitor.java:33]: Yeah, I can rename the parameters.

File: data-prepper-logstash-configuration/src/main/antlr/Logstash.g4 (hunk 2 above)
Comment (sshivanii) [reply data-prepper-logstash-configuration/src/main/antlr/Logstash.g4:1]: My bad, I meant adding a comment for what this grammar file consists of.

**Comment by sshivanii** [comment]
Sidenote: The first commit for this PR `e56cf3ee9fd2beb3e1234731d0b4b4749941f1d5` is a Merge commit. Based on the coding guidance we discussed, we're opting to Rebase instead of Merge to avoid any unnecessary Merge commits.

File: data-prepper-logstash-configuration/src/test/java/org/opensearch/dataprepper/logstash/parser/LogstashVisitorTest.java [hunk 6]
```diff
@@ -0,0 +1,559 @@
...
+        LogstashVisitor logstashVisitor = createObjectUnderTest();
+        Mockito.doReturn(TestDataProvider.pluginWithOneArrayContextAttributeData()).when(logstashVisitor).visitBranch_or_plugin(branchOrPluginContextMock);
+
+        LogstashConfiguration actualLogstashConfiguration = (LogstashConfiguration) logstashVisitor.visitConfig(configContextMock);
+        LogstashConfiguration expectedLogstashConfiguration = TestDataProvider.configData();
+
+        assertThat(actualLogstashConfiguration.getPluginSection(LogstashPluginType.INPUT).size(),
+                equalTo(expectedLogstashConfiguration.getPluginSection(LogstashPluginType.INPUT).size()));
+        Mockito.verify(logstashVisitor, Mockito.times(1)).visitPlugin_section(pluginSectionMock);
+    }
```
Comment (cmanning09) [inline data-prepper-logstash-configuration/src/test/java/org/opensearch/dataprepper/logstash/parser/LogstashVisitorTest.java:53]: It appears you are mocking the same objects over and over again in each test. I would encourage you to use `@Mock` annotation and create a class variable to reuse. I would also encourage you to create a LogstashVisitor class variable to assign in a `@BeforeEach` function as well. This will help reduce the duplicate code you have.

File: data-prepper-logstash-configuration/src/test/java/org/opensearch/dataprepper/logstash/parser/LogstashVisitorTest.java [hunk 7]
```diff
@@ -0,0 +1,559 @@
...
+import java.util.Map;
+
+class LogstashVisitorTest {
+
+    private LogstashVisitor createObjectUnderTest() {
+        return spy(new LogstashVisitor());
+    }
+
+    @Test
+    void visit_config_test() {
```
Comment (cmanning09) [inline data-prepper-logstash-configuration/src/test/java/org/opensearch/dataprepper/logstash/parser/LogstashVisitorTest.java:33]: This name is not very descriptive. What are we testing for here?

File: data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/parser/LogstashVisitor.java [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+import org.opensearch.dataprepper.logstash.model.LogstashValueType;
+
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.LinkedList;
+import java.util.ArrayList;
+
+/**
+ * Class to populate Logstash configuration model objects using ANTLR
```
Comment (cmanning09) [inline data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/parser/LogstashVisitor.java:20]: Can you add the following to the documentation? I am not familiar with ANTLR and how this whole process works. It took me a little while to understand where the `LogstashBaseVisitor` is coming from. I didn't realize I had to build my code to generate the class.
- What is a LogstashBaseVisitor? 
- How is a LogstashBaseVisitor generated?
- Where do all the methods come from?

**Review by cmanning09 — CHANGES_REQUESTED** [review]

File: data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/parser/LogstashVisitor.java (hunk 8 above)
Comment (dlvenable) [reply data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/parser/LogstashVisitor.java:20]: While this is confusing in a PR, this is all part of the build process and I think detailed documentation here is not quite appropriate. If anything, I'd keep it as simple as "... using ANTLR library classes and generated code."

File: data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/parser/LogstashVisitor.java [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+ *
+ * @since 1.2
+ */
+
```
Comment (dlvenable) [inline data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/parser/LogstashVisitor.java:24]: Please remove extra lines between Javadocs and their classes/methods. I found this one, but there may be others.

File: data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/parser/LogstashVisitor.java [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+@SuppressWarnings("rawtypes")
+public class LogstashVisitor extends LogstashBaseVisitor {
```
Comment (dlvenable) [inline data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/parser/LogstashVisitor.java:26]: This should be package private so that it is not used outside of this work.

File: data-prepper-logstash-configuration/src/test/java/org/opensearch/dataprepper/logstash/parser/TestDataProvider.java [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+import java.util.Collections;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.LinkedList;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+
+
+public class TestDataProvider {
```
Comment (dlvenable) [inline data-prepper-logstash-configuration/src/test/java/org/opensearch/dataprepper/logstash/parser/TestDataProvider.java:21]: This class should also be package private.

File: data-prepper-logstash-configuration/src/main/antlr/Logstash.g4 [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+    package org.opensearch.dataprepper.logstash;
+}
+/*
+* Parser Rules
+*/
+config: filler plugin_section filler (filler plugin_section)* filler;
+
+filler: (COMMENT | WS | NEWLINE)*;
+
+plugin_section: plugin_type filler '{'
```
[the remark this answers is not in the document]
Comment (dlvenable) [reply data-prepper-logstash-configuration/src/main/antlr/Logstash.g4:17]: I think this raises a question as to what convention we should use in the Grammar - follow the convention from Logstash, which would make it easier to map back to it; or use conventions which make the Java code nicer. This code should be used exclusively by the Logstash configuration framework, so I'm fine with using the Logstash conventions.

File: data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/exception/LogstashParsingException.java [hunk 13]
```diff
@@ -0,0 +1,14 @@
+package org.opensearch.dataprepper.logstash.exception;
+
+/**
+ * Exception thrown when {@link org.opensearch.dataprepper.logstash.parser.LogstashVisitor} is unable to convert
```
Comment (dlvenable) [inline data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/exception/LogstashParsingException.java:4]: I'm a big fan of `{@link}`, but I suggest removing it here. The `LogstashVisitor` is an internal implementation detail. Also, exceptions should generally not declare which classes use them. It goes the other way: classes specify which exceptions they throw.

I suggest rewording this somewhat.

**Review by dlvenable — CHANGES_REQUESTED** [review]

File: data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/parser/LogstashVisitor.java (hunk 8 above)
Comment (cmanning09) [reply data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/parser/LogstashVisitor.java:20]: That works. Anything to point the reader in the right direction.

File: data-prepper-logstash-configuration/src/main/antlr/Logstash.g4 (hunk 12 above)
[the remark this answers is not in the document]
Comment (cmanning09) [reply data-prepper-logstash-configuration/src/main/antlr/Logstash.g4:17]: Ahh okay. given this maps directly and is a manual process I am onboard with keep this as is to assist in the manual mapping process.  Thanks for clarifying.

File: data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/parser/LogstashVisitor.java [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+            throw new LogstashParsingException("conditionals are not supported");
+        }
+    }
+
+    @Override
+    public Object visitPlugin(LogstashParser.PluginContext pluginContext) {
+        String pluginName = pluginContext.name().getText();
+        List<LogstashAttribute> logstashAttributeList = new ArrayList<>();
+
+        for (int i = 0; i < pluginContext.attributes().attribute().size(); i++) {
```
[the remark this answers is not in the document]
Comment (asifsmohammed) [reply data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/parser/LogstashVisitor.java:87]: So the attributes will have list of attribute in grammar.

> attributes
> > attribute 1
> > attribute 2

We are storing each attribute from configuration in a list of type LogstashAttribute.

File: data-prepper-logstash-configuration/src/main/antlr/Logstash.g4 [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+    (
+        ('!' filler '(' filler condition filler ')')
+      | ('!' filler selector)
+    );
+
+in_expression: rvalue filler in_operator filler rvalue;
+
+not_in_expression: rvalue filler not_in_operator filler rvalue;
+
+rvalue: STRING | NUMBER | selector | array | method_call | regexp;
```
[the remark this answers is not in the document]
Comment (asifsmohammed) [reply data-prepper-logstash-configuration/src/main/antlr/Logstash.g4:95]: value is just value set for attributes. While rvalue is used in conditional statement.
We don't use rvalue currently, but we need to define all the rules in the grammar to parse the configuration.

**Review by dlvenable — DISMISSED** [review]

File: data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/parser/LogstashVisitor.java [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+/**
+ * Class to populate Logstash configuration model objects using ANTLR library classes and generated code.
+ *
+ * @since 1.2
+ */
+@SuppressWarnings("rawtypes")
+class LogstashVisitor extends LogstashBaseVisitor {
+
+    @Override
+    public Object visitConfig(LogstashParser.ConfigContext configContext) {
```
Comment (cmanning09) [inline data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/parser/LogstashVisitor.java:29]: We can mark every parameter for all methods implemented in this class as `final`

File: data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/parser/LogstashVisitor.java [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+        else if (attributeContext.value().BAREWORD() != null && attributeContext.value().getText().equals(attributeContext.value().BAREWORD().toString())) {
+            logstashValueType = LogstashValueType.BAREWORD;
+            value = attributeContext.value().getText();
+        }
+        else if (attributeContext.value().STRING() != null && attributeContext.value().getText().equals(attributeContext.value().STRING().toString())) {
+            logstashValueType = LogstashValueType.STRING;
+            value = attributeContext.value().getText().replaceAll("^\"|\"$|^'|'$", "");
+        }
+
+        LogstashAttributeValue logstashAttributeValue =  LogstashAttributeValue.builder()
```
Comment (cmanning09) [inline data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/parser/LogstashVisitor.java:105]: `final`

File: data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/exception/LogstashGrammarException.java [hunk 18]
```diff
@@ -0,0 +1,13 @@
+package org.opensearch.dataprepper.logstash.exception;
+
+/**
+ * Exception thrown when ANTLR fails to parse the Logstash configuration
+ *
+ * @since 1.2
+ */
+public class LogstashGrammarException extends LogstashParsingException {
```
Comment (cmanning09) [inline data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/exception/LogstashGrammarException.java:8]: This exception is never used. Will it be used in the future?

Comment (asifsmohammed) [reply data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/exception/LogstashGrammarException.java:8]: Yes. The intention was to use it in the future.

**Review by dlvenable — APPROVED** [review]

**Review by cmanning09 — APPROVED** [review]
